optionally set linux socket priority on DHCPv4 raw socket

Strangelovian opened the pull request:
Add new option [DHCPv4] SocketPriority.
Only if set, calls setsockopt(SOL_SOCKET, SO_PRIORITY, ... with the configured integer value; only for the DHCPv4 raw socket.

Use case: using systemd-networkd to implement internet home router, and ISP WAN interface impose 802.1Q VLAN ethernet frame priority set for received DHCPv4 packets.

Since DHCPv4 packets aren't alterable with netfilter (because of the raw socket), it  can be achieved with a setsockopt() call for option SO_PRIORITY.
Used with Netdev VLAN Egress QOS Maps option, this will send the needed VLAN priority for the (raw socket) DHCPv4 discover packets.

For UDP sockets, this is not needed since netfilter can mangle the needed priority.

Strangelovian commented:
Real life test OK on my debian 11 setup:
- network [DHCPv4] SocketPriority=6
- netdev [VLAN] EgressQOSMaps=6-6

Result: DHCP discover packet sent in ethernet VLAN frame with priority 6

poettering commented on src/libsystemd-network/dhcp-network.c:
```diff
@@ -113,6 +114,10 @@ static int _bind_raw_socket(
 
+        r = setsockopt_int(s, SOL_SOCKET, SO_PRIORITY, so_priority);
+        if (r < 0)
+                return -errno;
```
`return r;`

Strangelovian replied:
fixed in the squashed commit

poettering commented on src/network/networkd-dhcp4.c:
```diff
@@ -1499,6 +1499,12 @@ static int dhcp4_configure(Link *link) {
                         return log_link_debug_errno(link, r, "DHCPv4 CLIENT: Failed to set IP service type: %m");
         }
 
+        if (link->network->dhcp_socket_priority >= 0) {
```
hmm, are you sure only positive priorities are allowed? The way i read the kernel code no range check is done and negative priorities are ok too?

Strangelovian replied:
Yes you're right. Usual values are 0-6 but some people may use it for very specific stuff.
I removed the check.

poettering commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,15 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+        <listitem>
+          <para>An integer value assigned as the linux internal socket priority of the
+          raw IP socket used to send initial DHCPv4 messages. If not set, the default raw socket priority
+          is unchanged.</para>
```
must specify valid range and default

poettering replied:
should also mention that this exposes SO_PRIORITY plus a man page reference to where that's documented.

Strangelovian replied:
Fixed in the squashed commit.

poettering commented on NEWS:
```diff
@@ -48,6 +48,12 @@ CHANGES WITH 253 in spe:
         * systemd-fstab-generator now supports x-systemd.makefs option for
           /sysroot (in the initrd).
 
+        * New Network param SocketPriority to assign the Linux socket priority
```
suffix `SocketPriority=` with `=` please, as we usually do

Strangelovian replied:
Fixed in the squashed commit.

poettering commented:
also, please squah into a single commit with proper commit msg

Strangelovian commented:
@poettering Thanks for your review.
I made changes accordingly, and I forced update with a squashed commit.

poettering commented:
hmm, so hmm, with this change you'll now override the SO_PRIORITY value even if nothing was ever set.

I am pretty sure this is not what we want, because it means we'll undo any setting of that field via BPF programs. After all the priority is settable via bpf. (see struct bpf_sock).

This means by default we should not call SO_PRIORITY at all here.

Thus you need to make these two fields: the actual value, and a boolean that indicates whether it was set. And then only load the value via SO_PRIORITY if the boolean is true.

poettering commented:
an ebpf program attached to a programs cgroup.

i.e. similar to how "ip vrf exec" overrides SO_BINDTODEVICE on invoked processes. The same logic works for SO_PRIORITY as well, and we shouldn#t break that by overriding it always.

Strangelovian commented:
> an ebpf program attached to a programs cgroup.
> 
> i.e. similar to how "ip vrf exec" overrides SO_BINDTODEVICE on invoked processes. The same logic works for SO_PRIORITY as well, and we shouldn#t break that by overriding it always.

alright this clarifies, thanks

Strangelovian commented:
> Thus you need to make these two fields: the actual value, and a boolean that indicates whether it was set. And then only load the value via SO_PRIORITY if the boolean is true.

I added a conditional check to avoid overriding priority every time, please let me know about it:
- if priority not present in network file: no setsockopt SO_PRIORITY issued, no disruption of ebpf alterations
- if priority set to 0: not called either, but 0 is the default priority value

poettering commented:
> I added a conditional check to avoid overriding priority every time, please let me know about it:
> 
>     * if priority not present in network file: no setsockopt SO_PRIORITY issued, no disruption of ebpf alterations
> 
>     * if priority set to 0: not called either, but 0 is the default priority value

I think if people put SocketPriotity=0 in, then we should do the SO_PRIORITY call with 0.

Hence what I said earlier: this should be two fields internally: the priority integer *and* a boolean that indicates whether to actually apply it.

poettering commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -2213,6 +2223,7 @@ int sd_dhcp_client_new(sd_dhcp_client **ret, int anonymize) {
                 .anonymize = !!anonymize,
                 .max_attempts = UINT64_MAX,
                 .ip_service_type = -1,
+                .socket_priority = 0,
```
all fields that aren't listed are zero-initialized anyway, hence please drop this line

Strangelovian replied:
> Hence what I said earlier: this should be two fields internally: the priority integer _and_ a boolean that indicates whether to actually apply it.

Thanks, I added it in the very last updated commit, please let me know about it.
I also applied changes according to your other remarks.

poettering commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,18 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+      <varlistentry>
+        <term><varname>SocketPriority=</varname></term>
+        <listitem>
+          <para>The linux socket option SO_PRIORITY applied to the raw IP socket used for
```
please wrap `SO_PRIORITY` in `<constant>`

Strangelovian commented:
> Hence what I said earlier: this should be two fields internally: the priority integer _and_ a boolean that indicates whether to actually apply it.

Thanks, I added it in the very last updated commit, please let me know about it.
I also applied changes according to your other remarks.

poettering commented on src/libsystemd-network/dhcp-network.c:
```diff
@@ -113,6 +115,13 @@ static int _bind_raw_socket(
                 return -errno;
 
+        if (so_priority_set)
+        {
```
coding style, please put `{` on the same line as `if`, like we do everywhere else

poettering commented:
lgtm, just some nitpicks.

Will leave the final review/merge to @yuwata

Strangelovian commented:
> lgtm, just some nitpicks.
> Will leave the final review/merge to @yuwata
Thanks @poettering, your last remarks have been pushed.
@yuwata I'm looking forward to your review

Strangelovian commented:
> > lgtm, just some nitpicks.
> > Will leave the final review/merge to @yuwata

@yuwata hi sorry if I'm rude / too pushy... Can I request a bit of your time for reviewing it?

yuwata commented on NEWS:
```diff
@@ -48,6 +48,12 @@ CHANGES WITH 253 in spe:
         * systemd-fstab-generator now supports x-systemd.makefs option for
           /sysroot (in the initrd).
 
+        * New Network param SocketPriority= to assign the Linux socket priority
```
Please mention the section name, `[DHCPv4]`.

yuwata commented on NEWS:
```diff
@@ -48,6 +48,12 @@ CHANGES WITH 253 in spe:
 
+        * New Network param SocketPriority= to assign the Linux socket priority
+          used by DHCPv4 raw socket. Can be used in conjuntion with Link
+          VLAN EgressQosMappings to send the desired ethernet 802.1Q frame
```
Maybe better to mention which can be configured through `EgressQosMaps=` setting in `[VLAN]` section of .netdev file.

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+          <para>The linux socket option <constant>SO_PRIORITY</constant> applied to the raw IP socket used for
+          initial DHCPv4 messages. Defaults to 0. Usual values ranges from 0 to 6.
+          More details about <constant>SO_PRIORITY</constant> socket option in
+          <citerefentry><refentrytitle>socket</refentrytitle><manvolnum>7</manvolnum></citerefentry>
```
Missing period at the end.

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+          initial DHCPv4 messages. Defaults to 0. Usual values ranges from 0 to 6.
+          More details about <constant>SO_PRIORITY</constant> socket option in
+          <citerefentry><refentrytitle>socket</refentrytitle><manvolnum>7</manvolnum></citerefentry>
+          Can be used in conjunction with netdev VLAN EgressQOSMap to set the 802.1Q VLAN ethernet
```
`<varname>EgressQosMaps=</varname>`
As also said in the above, please clarify the section name `[VLAN]`.

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+          <citerefentry><refentrytitle>socket</refentrytitle><manvolnum>7</manvolnum></citerefentry>
+          Can be used in conjunction with netdev VLAN EgressQOSMap to set the 802.1Q VLAN ethernet
+          header priority, see
+          <citerefentry><refentrytitle>systemd.netdev</refentrytitle><manvolnum>5</manvolnum></citerefentry>
```
Missing period at the end.

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+        <term><varname>SocketPriority=</varname></term>
+        <listitem>
+          <para>The linux socket option <constant>SO_PRIORITY</constant> applied to the raw IP socket used for
+          initial DHCPv4 messages. Defaults to 0. Usual values ranges from 0 to 6.
```
IIUC the code, `Defaults to unset.`

yuwata commented on src/libsystemd-network/fuzz-dhcp-client.c:
```diff
@@ -16,7 +16,9 @@ int dhcp_network_bind_raw_socket(
                 const struct hw_addr_data *hw_addr,
                 const struct hw_addr_data *bcast_addr,
-                uint16_t arp_type, uint16_t port) {
+                uint16_t arp_type, uint16_t port,
```
Not your fault, but please break the line before `uint16_t port`. That is, one argument per line.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1701,6 +1707,43 @@ int config_parse_dhcp_ip_service_type(
+                void *userdata) {
+
+        Network *network = ASSERT_PTR(data);
+        int a;
+        int r;
```
Please merge the variable declarations into single line.

yuwata commented on src/network/networkd-network.h:
```diff
@@ -130,6 +130,8 @@ struct Network {
         uint16_t dhcp_client_port;
         int dhcp_critical;
         int dhcp_ip_service_type;
+        bool dhcp_socket_priority_set;
+        int dhcp_socket_priority;
```
Please swap the lines for better packing.

yuwata reviewed (changes requested):
Looks mostly good. Several minor comments about coding style and documentation.

Strangelovian commented:
> Looks mostly good. Several minor comments about coding style and documentation.
> Sorry for late review. I took slightly long new year's holidays. This is the first day of my work this year.

Thanks for reviewing and sorry for the aggressive chase up.
The last pushed commit addressed your remarks, please let me know about it.

yuwata commented on NEWS:
```diff
@@ -48,6 +48,12 @@ CHANGES WITH 253 in spe:
         * systemd-fstab-generator now supports x-systemd.makefs option for
           /sysroot (in the initrd).
 
+        * New Network param [DHCPv4]SocketPriority= to assign the Linux socket priority
```
`New Network param [DHCPv4]SocketPriority=` -> `The [DHCPv4] section in .network file gained new SocketPriority= setting that assigns the Linux socket priority...`

yuwata commented on NEWS:
```diff
@@ -48,6 +48,12 @@ CHANGES WITH 253 in spe:
           /sysroot (in the initrd).
 
+        * New Network param [DHCPv4]SocketPriority= to assign the Linux socket priority
+          used by the DHCPv4 raw socket. Can be used in conjuntion with the EgressQosMaps=
```
`EgressQosMaps=` -> `EgressQOSMaps=`

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+      <varlistentry>
+        <term><varname>SocketPriority=</varname></term>
+        <listitem>
+          <para>The linux socket option <constant>SO_PRIORITY</constant> applied to the raw IP socket used for
```
`linux` -> `Linux`

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1701,6 +1707,42 @@ int config_parse_dhcp_ip_service_type(
+                void *userdata) {
+
+        Network *network = ASSERT_PTR(data);
+        int a,r;
```
Please insert a space between `a` and `r`: `int a, r;`

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+          <para>The linux socket option <constant>SO_PRIORITY</constant> applied to the raw IP socket used for
+          initial DHCPv4 messages. Unset by default. Usual values range from 0 to 6.
+          More details about <constant>SO_PRIORITY</constant> socket option in
+          <citerefentry><refentrytitle>socket</refentrytitle><manvolnum>7</manvolnum></citerefentry>.
```
`<citerefentry project='man-pages'>`, I guess.

Strangelovian commented:
@yuwata Thanks, I pushed a new commit with all your last batch of remarks

yuwata commented on man/systemd.network.xml:
```diff
@@ -1890,6 +1890,20 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+          <citerefentry project='man-pages'><refentrytitle>socket</refentrytitle><manvolnum>7</manvolnum></citerefentry>.
+          Can be used in conjunction with [VLAN] section <varname>EgressQOSMaps=</varname> setting of .netdev
+          file to set the 802.1Q VLAN ethernet tagged header priority, see
+          <citerefentry project='man-pages'><refentrytitle>systemd.netdev</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
```
Not familiar, but looking at the other citerefentry tags, project= should be dropped in this case.

Strangelovian replied:
Alright, pushed a new commit with this

Strangelovian commented:
@yutawa I addressed your last remarks about man page, please let me know about it
